File: Examples/HistoricalDataRequest.cs
//------------------------------------------------------------------------------
// <copyright project="Examples" file="HistoricalDataRequest.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Examples
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    using BEmu; //un-comment this line to use the Bloomberg API Emulator
    //using Bloomberglp.Blpapi; //un-comment this line to use the actual Bloomberg API

    public static class HistoricalDataRequest
    {
        public static void RunExample()
        {
            SessionOptions sessionOptions = new SessionOptions();
            sessionOptions.ServerHost = "127.0.0.1";
            sessionOptions.ServerPort = 8194;

            Session session = new Session(sessionOptions);
            session.Start();
            session.OpenService("//blp/refdata");

            Service service = session.GetService("//blp/refdata");

            Request request = service.CreateRequest("HistoricalDataRequest");

            //request information for the following securities
            request.Append("securities", "MSFT US EQUITY");
            //request.Append("securities", "ZYZZ US EQUITY"); //the code treats securities that start with a "Z" as non-existent
            request.Append("securities", "C A COMDTY");
            request.Append("securities", "AAPL 150117C00600000 EQUITY"); //this is a stock option: TICKER yyMMdd[C/P]\d{8} EQUITY

            //include the following simple fields in the result
            //request.Append("fields", "ZBID"); //the code treats a field that starts with a "Z" as a bad field
            request.Append("fields", "BID");
            request.Append("fields", "ASK");

            //Historical requests allow a few overrides.  See the developer's guide A.2.4 for more information.

            request.Set("startDate", DateTime.Today.AddMonths(-1).ToString("yyyyMMdd")); //Request that the information start three months ago from today.  This override is required.
            request.Set("endDate", DateTime.Today.AddDays(10).ToString("yyyyMMdd")); //Request that the information end three days before today.  This is an optional override.  The default is today.
            
            //Determine the frequency and calendar type of the output. To be used in conjunction with Period Selection.
            request.Set("periodicityAdjustment", "CALENDAR"); //Optional string.  Valid values are ACTUAL (default), CALENDAR, and FISCAL.

            //Determine the frequency of the output. To be used in conjunction with Period Adjustment.
            request.Set("periodicitySelection", "DAILY"); //Optional string.  Valid values are DAILY (default), WEEKLY, MONTHLY, QUARTERLY, SEMI_ANNUALLY, and YEARLY

            //Sets quote to Price or Yield for a debt instrument whose default value is quoted in yield (depending on pricing source).
            request.Set("pricingOption", "PRICING_OPTION_PRICE"); //Optional string.  Valid values are PRICING_OPTION_PRICE (default) and PRICING_OPTION_YIELD

            //Adjust for "change on day"
            request.Set("adjustmentNormal", true); //Optional bool. Valid values are true and false (default = false)

            //Adjusts for Anormal Cash Dividends
            request.Set("adjustmentAbnormal", false); //Optional bool. Valid values are true and false (default = false)

            //Capital Changes Defaults
            request.Set("adjustmentSplit", true); //Optional bool. Valid values are true and false (default = false)

            //The maximum number of data points to return, starting from the startDate
            //request.Set("maxDataPoints", 5); //Optional integer.  Valid values are positive integers.  The default is unspecified in which case the response will have all data points between startDate and endDate

            //Indicates whether to use the average or the closing price in quote calculation.
            request.Set("overrideOption", "OVERRIDE_OPTION_CLOSE"); //Optional string.  Valid values are OVERRIDE_OPTION_GPA for an average and OVERRIDE_OPTION_CLOSE (default) for the closing price

            CorrelationID requestID = new CorrelationID(1);
            session.SendRequest(request, requestID);

            bool continueToLoop = true;
            while (continueToLoop)
            {
                Event eventObj = session.NextEvent();
                switch (eventObj.Type)
                {
                    case Event.EventType.RESPONSE: // final event
                        continueToLoop = false;
                        handleResponseEvent(eventObj);
                        break;
                    case Event.EventType.PARTIAL_RESPONSE:
                        handleResponseEvent(eventObj);
                        break;
                    default:
                        handleOtherEvent(eventObj);
                        break;
                }
            }
        }

        private static void handleResponseEvent(Event eventObj)
        {
            Console.WriteLine("EventType = " + eventObj.Type);
            foreach (Message message in eventObj.GetMessages())
            {
                Console.WriteLine();
                Console.WriteLine("correlationID= " + message.CorrelationID);
                Console.WriteLine("messageType = " + message.MessageType);
                
                Element elmSecurityData = message["securityData"];

                Element elmSecurity = elmSecurityData["security"];
                string security = elmSecurity.GetValueAsString();
                Console.WriteLine(security);

                if (elmSecurityData.HasElement("securityError", true))
                {
                    Element elmSecError = elmSecurityData["securityError"];
                    string source = elmSecError.GetElementAsString("source");
                    int code = elmSecError.GetElementAsInt32("code");
                    string category = elmSecError.GetElementAsString("category");
                    string errorMessage = elmSecError.GetElementAsString("message");
                    string subCategory = elmSecError.GetElementAsString("subcategory");

                    Console.WriteLine("security error");
                    Console.WriteLine(string.Format("source = {0}", source));
                    Console.WriteLine(string.Format("code = {0}", code));
                    Console.WriteLine(string.Format("category = {0}", category));
                    Console.WriteLine(string.Format("errorMessage = {0}", errorMessage));
                    Console.WriteLine(string.Format("subCategory = {0}", subCategory));
                }
                else
                {
                    bool hasFieldErrors = elmSecurityData.HasElement("fieldExceptions", true);
                    if (hasFieldErrors)
                    {
                        Element elmFieldErrors = elmSecurityData["fieldExceptions"];
                        for (int errorIndex = 0; errorIndex < elmFieldErrors.NumValues; errorIndex++)
                        {
                            Element fieldError = elmFieldErrors.GetValueAsElement(errorIndex);
                            string fieldId = fieldError.GetElementAsString("fieldId");

                            Element errorInfo = fieldError["errorInfo"];
                            string source = errorInfo.GetElementAsString("source");
                            int code = errorInfo.GetElementAsInt32("code");
                            string category = errorInfo.GetElementAsString("category");
                            string strMessage = errorInfo.GetElementAsString("message");
                            string subCategory = errorInfo.GetElementAsString("subcategory");

                            Console.WriteLine();
                            Console.WriteLine();
                            Console.WriteLine("field error: ");
                            Console.WriteLine(string.Format("\tfieldId = {0}", fieldId));
                            Console.WriteLine(string.Format("\tsource = {0}", source));
                            Console.WriteLine(string.Format("\tcode = {0}", code));
                            Console.WriteLine(string.Format("\tcategory = {0}", category));
                            Console.WriteLine(string.Format("\terrorMessage = {0}", strMessage));
                            Console.WriteLine(string.Format("\tsubCategory = {0}", subCategory));
                        }
                    }

                    Element elmFieldData = elmSecurityData["fieldData"];
                    for (int valueIndex = 0; valueIndex < elmFieldData.NumValues; valueIndex++)
                    {
                        Element elmValues = elmFieldData.GetValueAsElement(valueIndex);
                        DateTime date = elmValues.GetElementAsDate("date").ToSystemDateTime();
                        double bid = elmValues.GetElementAsFloat64("BID");
                        double ask = elmValues.GetElementAsFloat64("ASK");

                        Console.WriteLine(string.Format("{0:yyyy-MM-dd}: BID = {1}, ASK = {2}", date, bid, ask));
                    }
                }
            }
        }

        private static void handleOtherEvent(Event eventObj)
        {
            Console.WriteLine("EventType=" + eventObj.Type);
            foreach (Message message in eventObj.GetMessages())
            {
                Console.WriteLine("correlationID=" + message.CorrelationID);
                Console.WriteLine("messageType=" + message.MessageType);
                Console.WriteLine(message.ToString());
                if (Event.EventType.SESSION_STATUS == eventObj.Type && message.MessageType.Equals("SessionTerminated"))
                {
                    Console.WriteLine("Terminating: " + message.MessageType);
                }
            }
        }
    }
}

File: Examples_cpp/src/HistoricalDataRequest.cpp
//------------------------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/HistoricalDataRequest.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------------------------

#include <iostream>
#include <sstream>
#include <time.h>
#include "api_selector.h"

using namespace std;

namespace Examples
{
	void ProcessHistoricalDataEvent(Event evt);
	char * DatetimeToYYYYMMDD(const Datetime& arg);

	void RunHistoricalDataExample()
	{
		SessionOptions soptions;
		soptions.setServerHost("localhost");
		soptions.setServerPort(8194);

		Session session(soptions);
		if(session.start() && session.openService("//blp/refdata"))
		{
			Service refDataSvc = session.getService("//blp/refdata");

			Request request = refDataSvc.createRequest("HistoricalDataRequest");

            //request information for the following securities
            request.append("securities", "MSFT US EQUITY");
            request.append("securities", "C A COMDTY");
            request.append("securities", "AAPL 150117C00600000 EQUITY"); //this is a stock option: TICKER yyMMdd[C/P]\d{8} EQUITY

            //uncomment the following line to see what a request for a nonexistent security looks like
            //request.append("securities", "ZIBM US EQUITY");
            //  My code treats all securities that start with a 'Z' as a nonexistent security


            //include the following simple fields in the result
            request.append("fields", "BID");
            request.append("fields", "ASK");

            //uncomment the following line to see what a request for an invalid field looks like
            //request.append("fields", "ZBID");
            //  My code treats all fields that start with a 'Z' as an invalid field

            //Historical requests allow a few overrides.  See the developer's guide A.2.4 for more information.
			
			//30 days in the past
			time_t tStart = time(NULL) - (30 * 24 * 60 * 60); //days * hours * minutes * seconds
			struct tm *aTimeStart = localtime(&tStart); //Visual Studio complains about this call, but the localtime_s replacement is not portable.
			int day = aTimeStart->tm_mday;
			int month = aTimeStart->tm_mon + 1;
			int year = aTimeStart->tm_year + 1900;
			Datetime dtStart(year, month, day);
			char * strDtStart = DatetimeToYYYYMMDD(dtStart);
			request.set("startDate", strDtStart); //Request that the information start three months ago from today.  This override is required.
			delete strDtStart;
			strDtStart = 0;

			//10 days in the future
			time_t tEnd = time(NULL) + (10 * 24 * 60 * 60); //days * hours * minutes * seconds
			struct tm *aTimeEnd = localtime(&tEnd); //Visual Studio complains about this call, but the localtime_s replacement is not portable.
			day = aTimeEnd->tm_mday;
			month = aTimeEnd->tm_mon + 1;
			year = aTimeEnd->tm_year + 1900;
			Datetime dtEnd(year, month, day);
			char * strDtEnd = DatetimeToYYYYMMDD(dtEnd);
            request.set("endDate", strDtEnd); //Request that the information end three days before today.  This is an optional override.  The default is today.
			delete strDtEnd;
			strDtEnd = 0;


            //Determine the frequency and calendar type of the output. To be used in conjunction with Period Selection.
            request.set("periodicityAdjustment", "CALENDAR"); //Optional string.  Valid values are ACTUAL (default), CALENDAR, and FISCAL.

            //Determine the frequency of the output. To be used in conjunction with Period Adjustment.
            request.set("periodicitySelection", "DAILY"); //Optional string.  Valid values are DAILY (default), WEEKLY, MONTHLY, QUARTERLY, SEMI_ANNUALLY, and YEARLY

            //Sets quote to Price or Yield for a debt instrument whose default value is quoted in yield (depending on pricing source).
            request.set("pricingOption", "PRICING_OPTION_PRICE"); //Optional string.  Valid values are PRICING_OPTION_PRICE (default) and PRICING_OPTION_YIELD

            //Adjust for "change on day"
            request.set("adjustmentNormal", true); //Optional bool. Valid values are true and false (default = false)

            //Adjusts for Anormal Cash Dividends
            request.set("adjustmentAbnormal", false); //Optional bool. Valid values are true and false (default = false)

            //Capital Changes Defaults
            request.set("adjustmentSplit", true); //Optional bool. Valid values are true and false (default = false)
			
            //The maximum number of data points to return, starting from the startDate
            //request.Set("maxDataPoints", 5); //Optional integer.  Valid values are positive integers.  The default is unspecified in which case the response will have all data points between startDate and endDate

            //Indicates whether to use the average or the closing price in quote calculation.
            request.set("overrideOption", "OVERRIDE_OPTION_CLOSE"); //Optional string.  Valid values are OVERRIDE_OPTION_GPA for an average and OVERRIDE_OPTION_CLOSE (default) for the closing price
			
			CorrelationId requestId(17); //17 is just some number I picked
			session.sendRequest(request, requestId);

			bool continueToLoop = true;
			while(continueToLoop)
			{
				Event evt(session.nextEvent());
				switch (evt.eventType())
				{
					case Event::RESPONSE:
						continueToLoop = false; //fall through
					case Event::PARTIAL_RESPONSE:
						ProcessHistoricalDataEvent(evt);
						break;
				}
			}

			session.stop();
		}
		else
		{
			cerr << "Cannot connect to server";
		}
	}

	void ProcessHistoricalDataEvent(Event evt)
	{
		cout << "EventType = " << evt.eventType();
		MessageIterator iter(evt);
		while(iter.next())
		{
			Message msg = iter.message();

			cout << endl;
			cout << "correlationID = " << msg.correlationId() << endl;
			cout << "messageType = " << msg.messageType() << endl;

			if(msg.hasElement("securityData", true))
			{
				Element elmSecurityData = msg.getElement("securityData");

				Element elmSecurity = elmSecurityData.getElement("security");
				const char * security = elmSecurity.getValueAsString();
				cout << security << endl;

				bool isSecurityError = elmSecurityData.hasElement("securityError", true);
				if(isSecurityError)
				{
					Element elmSecError = elmSecurityData.getElement("securityError");
					string source = elmSecError.getElementAsString("source");
					int code = elmSecError.getElementAsInt32("code");
					string category = elmSecError.getElementAsString("category");
					string errorMessage = elmSecError.getElementAsString("message");
					string subCategory = elmSecError.getElementAsString("subcategory");

					cerr << "security error" << endl;
					cerr << "source = " << source << endl;
					cerr << "code = " << code << endl;
					cerr << "category = " << category << endl;
					cerr << "errorMessage = " << errorMessage << endl;
					cerr << "subcategory = " << subCategory << endl;
				}
				else
				{
					bool hasFieldErrors = elmSecurityData.hasElement("fieldExceptions", true);
					if(hasFieldErrors)
					{
						Element elmFieldErrors = elmSecurityData.getElement("fieldExceptions");
						for (size_t errorIndex = 0; errorIndex < elmFieldErrors.numValues(); errorIndex++)
						{
							Element fieldError = elmFieldErrors.getValueAsElement(errorIndex);
							string fieldId = fieldError.getElementAsString("fieldId");

							Element errorInfo = fieldError.getElement("errorInfo");
							string source = errorInfo.getElementAsString("source");
							int code = errorInfo.getElementAsInt32("code");
							string category = errorInfo.getElementAsString("category");
							string strMessage = errorInfo.getElementAsString("message");
							string subCategory = errorInfo.getElementAsString("subcategory");

							cerr << endl;
							cerr << endl;
							cerr << "field error: " << endl;
							cerr << "\tfieldId = " << fieldId << endl;
							cerr << "\tsource = " << source << endl;
							cerr << "\tcode = " << code << endl;
							cerr << "\tcategory = " << category << endl;
							cerr << "\terrorMessage = " << strMessage << endl;
							cerr << "\tsubCategory = " << subCategory << endl;
						}
					}

					Element elmFieldData = elmSecurityData.getElement("fieldData");
					for (size_t valueIndex = 0; valueIndex < elmFieldData.numValues(); valueIndex++)
					{
						Element elmValues = elmFieldData.getValueAsElement(valueIndex);
						Datetime date = elmValues.getElementAsDatetime("date");
						double bid = elmValues.getElementAsFloat64("BID");
						double ask = elmValues.getElementAsFloat64("ASK");

						cout << date << ": BID = " << bid << ", ASK = " << ask << endl;
					}

				}
			}
			else
			{
				cout << "no security data" << endl;
				msg.print(cout);
			}

		}
	}

	char * DatetimeToYYYYMMDD(const Datetime& arg)
	{
		stringstream ss;

		ss << arg.year();

		if(arg.month() < 10)
			ss << '0';
		ss << arg.month();

		if(arg.day() < 10)
			ss << '0';
		ss << arg.day();

		string str = ss.str();
		const char * temp = str.c_str();

		char * result = new char[9];
		strncpy(result, temp, 9);

		return result;
	}

}